### PR TITLE
feat: Maintaining audit logs

### LINKED ISSUE #3665: Feature: Basic Audit logs on Devtron UI
### 🔖 Feature description

As an avid user of Devtron, I would like to propose the implementation of audit logs as a valuable feature for the platform. Audit logs are an essential aspect of any platform engineering tool, and their integration within Devtron would significantly elevate its functionality and security.

I believe that the addition of audit logs aligns perfectly with the platform's commitment to providing a robust, secure, and user-friendly experience. This feature will undoubtedly boost the confidence of the Devtron community and enterprise users and contribute to the platform's overall success.

### 🎤 Pitch

Importance of Audit logs:

**Security and Compliance:** With the ever-increasing need for robust security measures, having comprehensive audit logs in Devtron is crucial for detecting and investigating potential security breaches or unauthorized access attempts. This feature would enable us to uphold compliance standards and demonstrate adherence to data privacy regulations.

**User Accountability and Transparency:** Enhanced audit logs would empower us to track and monitor user activities within Devtron, attributing specific actions to individual users. This increased transparency fosters a sense of accountability and promotes responsible usage of the platform.

**Monitoring and Troubleshooting:** The ability to review detailed logs of activities and events within Devtron would aid in monitoring the platform's usage, identifying potential bottlenecks, and troubleshooting issues efficiently.

**Proactive Issue Detection:** By regularly reviewing the audit logs, we can proactively identify and address recurring errors or potential issues before they escalate, thus ensuring the seamless operation of deployed microservices.

### ✌️ Solution

The basic audit logs will only have a few parameters, such as:

| Parameters | Definition | Example |
| --- | --- | -- |
| action: | Action Performed | updateCIpipeline |  
| updated-by | User ID of user who took the action |  2 |
| updated-on | Time stamp of action | 2023-05-03 21:55:6 |
| api-request | API request without auth | Entire api curl without auth data |
| api-response-code | API response code/status | 200 |

1. Make this event logging mandatory at each update action triggered by a user or an api-user.
2. Provide filters for all the 3 fields above to filter and search relevant logs.

### 🔄️ Alternative

Users have to rely on getting the audit logs from multiple DB tables.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#4480

### PR DESCRIPTION
**Pull request by dhananjay0601**
# Description

Maintaining Audit log for each API call so that it would significantly elevate its functionality and security.

Fixes #3665 

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [X] I have performed a self-review of my own code.
* [X] I have commented my code, particularly in hard-to-understand areas.
* [X] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/logger/loggerService.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package logger
+
+import (
+	"log"
+	"time"
+)
+
+func NewUserAuthService(action string, updatedBy int32, updatedOn time.Duration, apiResponseCode int) {
+	log.Println(action)
```
Comment (vikramdevtron) [inline api/logger/loggerService.go:9]: use logger to write logs in single row

File: api/dashboardEvent/DashboardTelemetryRouter.go [hunk 2]
```diff
@@ -1,6 +1,8 @@
 package dashboardEvent
 
-import "github.com/gorilla/mux"
+import (
```
Comment (nishant-d) [inline api/dashboardEvent/DashboardTelemetryRouter.go:3]: unused import

File: api/logger/logger.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		startTime := time.Now()
+
+		// Call the next handler in the chain.
+		next.ServeHTTP(w, r)
+
+		//Log the request details.
+		url := r.URL.Path
+
+		token := r.Header.Get("token")
+		userId, userType, err := impl.userService.GetUserByToken(r.Context(), token)
```
Comment (nishant-d) [inline api/logger/logger.go:38]: log email instead of userID, its a db call.

File: api/logger/logger.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			log.Printf("userId does not exists")
+		}
+		vars := r.URL.Query().Encode()
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Printf("error reading request body")
+		}
+		// Convert the request body to a string
+		requestPayload := string(body)
```
Comment (nishant-d) [inline api/logger/logger.go:48]: avoid string conversion from performance POV

File: api/logger/logger.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+		token := r.Header.Get("token")
+		userId, userType, err := impl.userService.GetUserByToken(r.Context(), token)
+		if err != nil {
+			log.Printf("userId does not exists")
+		}
+		vars := r.URL.Query().Encode()
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Printf("error reading request body")
```
Comment (nishant-d) [inline api/logger/logger.go:45]: add some identifier here

File: api/logger/logger.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		userService: userService,
+	}
+}
+
+type LoggingMiddleware interface {
+	LoggingMiddleware(next http.Handler) http.Handler
+}
+
+// LoggingMiddleware is a middleware function that logs the incoming request.
+func (impl LoggingMiddlewareImpl) LoggingMiddleware(next http.Handler) http.Handler {
```
Comment (nishant-d) [inline api/logger/logger.go:26]: make this function super efficient.

**Review by nishant-d — CHANGES_REQUESTED** [review]

**Review by vikramdevtron — APPROVED** [review]

**Review by vikramdevtron — DISMISSED** [review]

**Review by vikramdevtron — APPROVED** [review]

**Review by nishant-d — APPROVED** [review]

**Pull request merged**